src/main/java/com/es/phoneshop/model/product/recentlyViewed/HttpSessionRecentlyViewedService.java

```diff
@@ -0,0 +1,54 @@
+package com.es.phoneshop.model.product.recentlyViewed;
+
+import com.es.phoneshop.model.product.ArrayListProductDao;
+import com.es.phoneshop.model.product.Product;
+import com.es.phoneshop.model.product.exceptions.ProductNotFoundException;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpSession;
+import java.util.ArrayList;
+import java.util.List;
+
+public class HttpSessionRecentlyViewedService implements RecentlyViewedService {
+    protected static final String HISTORY_SESSION_ATTRIBUTE = HttpSessionRecentlyViewedService.class + ".recentlyViewed";
+    private static final int MAX_RECENTLY_VIEWED_SIZE = 3;
+    private ArrayListProductDao arrayListProductDao = ArrayListProductDao.getInstance();
+
+    public static HttpSessionRecentlyViewedService getInstance() {
+        return SingletonHolder.instance;
+    }
+
+    public static class SingletonHolder {
+        public static final HttpSessionRecentlyViewedService instance = new HttpSessionRecentlyViewedService();
+    }
+
+    public void addRecentlyViewedProduct(List<Product> recentlyViewedProducts, Long productId) throws ProductNotFoundException {
+        Product product = arrayListProductDao.getProduct(productId);
+        int i;
+        for (i = 0; i < recentlyViewedProducts.size(); i++) {
```

Comment:
Why is it this complicated? Let's say we have a list. When a new product is visited:
- If it's already present - remove it (one line of code)
- Add this product to the head (one line of code)
- If the list size now exceeds 3 => remove the tail (3 lines of code)